### ISSUE TITLE
Do `struct rpma_completion` require rework?

### ISSUE THREAD

**Issue by janekmi** [pmem/rpma#976]
In the light of #958 it might be required to add yet another field to `struct rpma_completion` which would allow to track down the connection from which the completion comes from. Since after making CQ for receives shared by multiple connections it is no longer obvious from which connection the completion comes from.

It is another situation where this public structure might be extended (you might recall an immediate data extension a few months ago). It is not a good sign since, we hope, we are heading to the first stable release of the library. It might be a necessity to hide this structure internals and allow accessing its values only via getters but it would require allocating at runtime which is a drawback.

A possible rework might also include reconsideration of using `enum rpma_op`. At this point, it is a simple translation of `enum ibv_wc_opcode` which may be unnecessary.

To sum up, all things to consider:
- extending `struct rpma_completion` with a connection which the completion comes from
- banning direct access to `struct rpma_completion` fields and provide getters/setters
- the value of having `enum rpma_op`

**Comment by yangx-jy** [comment]
Hi @janekmi

> extending struct rpma_completion with a connection which the completion comes from
We can talk it on #977.

> banning direct access to struct rpma_completion fields and provide getters/setters
Is it necessary to ban direct access to struct rpma_completion fields？ I am worried about that it becomes complex to access all members in struct rpma_completion when struct rpma_completion has a lot of  members.

> the value of having enum rpma_op
Agreed. I have tried to simplify the conversion from (enum ibv_wc_opcode) opcode to (enum rpma_op) op by PR #979.

Best Regards,
Xiao Yang

**Comment by janekmi** [comment]
> Is it necessary to ban direct access to struct rpma_completion fields？ I am worried about that it becomes complex to access all members in struct rpma_completion when struct rpma_completion has a lot of members.

I know it would be less convenient. But if we would agree that changes to this structure in the future might be necessary then we can not keep it open. We have to consider both scenarios carefully.

> Agreed. I have tried to simplify the conversion from (enum ibv_wc_opcode) opcode to (enum rpma_op) op by PR #979.

Thanks for the effort. I think if we won't do any advanced conversion we should drop the `enum rpma_op` entirely since it introduces no value. So the completion structure would look as follows:

```c
struct rpma_completion {
	void *op_context;
	enum ibv_wc_opcode opcode; /* instead of enum rpma_op op */
	uint32_t byte_len;
	enum ibv_wc_status op_status;
	unsigned flags;
	uint32_t imm;
};
```

Additionally, I think we should put into the documentation what `enum ibv_wc_opcode` values may one expect when using the librpma API to provide guidelines for the developers.

## Alternatively

We can discuss finding a way to have a consistent `enum rpma_op` output no matter what ibverbs are in use e.g. translating `IBV_WC_RDMA_READ` to `RPMA_OP_FLUSH` when the `IBV_WC_RDMA_READ` completion comes from posted `rpma_flush()` operation. **Note**: The solution has to take into account possible user-provided `op_context` and posting `rpma_read()` at the same time.

**Comment by yangx-jy** [comment]
> Additionally, I think we should put into the documentation what `enum ibv_wc_opcode` values may one expect when using the librpma API to provide guidelines for the developers.

Hi @janekmi 

1)  If `enum ibv_wc_opcode` is used directly, it is not clear for user to know which rpma operation is completed and all existing examples/upper applications have to be rewritten. 
2）If `enum ibv_wc_opcode` is used directly, RPMA has no chance to report RPMA_OP_FLUSH currently until RDMA has been extended to support FLUSH operation(e.g. IBV_WC_RDMA_FLUSH).

> ## Alternatively
> 
> We can discuss finding a way to have a consistent `enum rpma_op` output no matter what ibverbs are in use e.g. translating `IBV_WC_RDMA_READ` to `RPMA_OP_FLUSH` when the `IBV_WC_RDMA_READ` completion comes from posted `rpma_flush()` operation. **Note**: The solution has to take into account possible user-provided `op_context` and posting `rpma_read()` at the same time.

Yes, we need to report RPMA_OP_FLUSH properly.  I will try this method.

Best Regards,
Xiao Yang

**Comment by janekmi** [comment]
> 1. If enum ibv_wc_opcode is used directly, it is not clear for user to know which rpma operation is completed and all existing examples/upper applications have to be rewritten.

Yes. It is a tough decision. But keeping a no-value function in the library is unbearable for me. ;-)

> 2. If enum ibv_wc_opcode is used directly, RPMA has no chance to report RPMA_OP_FLUSH currently until RDMA has been extended to support FLUSH operation(e.g. IBV_WC_RDMA_FLUSH).

Having `RPMA_OP_*` we are not doing it now either. This is why we are looking for a way for providing this.

> Yes, it is necessary to report RPMA_OP_FLUSH properly but hard to distinguish between read and flush completion events.

A 'hard' is what keeps us running. :-)

**Comment by janekmi** [comment]
In the face of upcoming #1087 the possibility of changing the `struct rpma_completion` is not theoretical anymore. Considering that:

- `struct rpma_completion` does not provide the initially assumed value and
- the upcoming changes revolts around transferring more and more values from `struct ibv_wc` to `struct rpma_completion` where
- extending `struct rpma_completion` breaks backwards compatibility and may introduce hard to debug issues in the future

the easiest way of addressing both these issues is simply **replacing `struct rpma_completion` with `struct ibv_wc`**.

# Idea 1

```c
/* deprecate */
int rpma_cq_get_completion(struct rpma_cq *cq, struct rpma_completion *cmpl);

int rpma_cq_get_wc(struct rpma_cq *cq, struct ibv_wc *wc);

int rpma_wc_get_imm(struct ibv_wc *wc, uint32_t *imm);
```

**Comment by janekmi** [comment]
> Do you want to remove it or keep it in next release?

No. I think it should be removed.

> I hope we can find a way to achieve its original value (i.e. report RPMA_OP_FLUSH -- hide the underlying read/flush operation)

We have already discussed the **Idea 2** many times internally. Sadly, it is the only light way of achieving the `RPMA_OP_FLUSH` purpose of existence. But I think it is not worth it. Currently, the op_context size exactly matches the size of a pointer which allows easily querying an unlimited number of information related to the operation. Whereas, when we decide to reserve a single bit out of the op_context you have to build some kind of dictionary or whatnot. The same result can be easily achieved by the end-user without compromising the fundamental role of the op_context which is fully reserved to the end-user.

TLDR: If it can be achieved by the end-user without the drawbacks we would introduce if we decide to help him with this matter I do not see a reason to do it.

> int rpma_wc_get_imm(struct ibv_wc *wc, uint32_t *imm);
>
> I think application can access all members of struct ibv_wc so is the function rpma_wc_get_imm necessary?

To-be-discussed. The same result the end-user can achieve simply by using the `ntohl()` directly. Which maybe we should just recommend.

**Comment by yangx-jy** [comment]
> TLDR: If it can be achieved by the end-user without the drawbacks we would introduce if we decide to help him with this matter I do not see a reason to do it.

Thanks for your detailed explanation.  In my opinion, librpma instead of end-user is responsible for distinguishing rpma flush and rpma read.  Maybe I'm wrong.

>To-be-discussed. The same result the end-user can achieve simply by using the ntohl() directly. Which maybe we should just recommend.

Agreed.  It is better to leave it to end-user.

**Comment by yangx-jy** [comment]
Hi @grom72 @ldorau 

I think you still perfer idea1(replace struct rpma_completion with `struct ibv_wc`). Right?
If right, I will update my PR in this way.

**Comment by yangx-jy** [comment]
Hi @grom72 @ldorau 

Let's look at the latest PR #979

**Comment by Patryk717** [comment]
[done](https://reviewable.io/reviews/pmem/rpma/1080)
[done](https://github.com/pmem/rpma/pull/979)
